evaluate_performance.py

```diff
@@ -0,0 +1,130 @@
+import json
+
+from functools import partial
+from sklearn.metrics import (
+    accuracy_score,
+    precision_score,
+    recall_score,
+    f1_score,
+    confusion_matrix,
+)
+from run_llm.benchmark_set_up import Benchmark, get_benchmark_by_name
+
+# Configuration
+# =============
+PATH = "results.json"  # Path to the results file
+PRINT_ALL = False  # Print all groups or just the top 5
+INCLUDE_ALL_METRICS_IN_PROPERTY = False  # Include all metrics in the property ranking
+RANK_BY = "f1"  # This is due to the unbalanced data. Can be: "accuracy", "precision", "recall", "f1"
+MIN_SAMPLES = 3  # Minimum number of samples to consider a group
+SAVE_RESULT_PATH = (
+    None  # If saving to file set it to something like: "eval_results.txt"
+)
+OUTPUT_FILE = open(SAVE_RESULT_PATH, "w") if SAVE_RESULT_PATH else None
+
+printo = partial(print, file=OUTPUT_FILE)
+
+
+# Functions
+# =========
+def calculate_metrics(ground_truths: list[str], predictions: list[str]):
+    return {
+        "accuracy": accuracy_score(ground_truths, predictions),
+        "precision": precision_score(
+            ground_truths, predictions, average="macro", zero_division=0
+        ),
+        "recall": recall_score(
+            ground_truths, predictions, average="macro", zero_division=0
+        ),
+        "f1": f1_score(ground_truths, predictions, average="macro"),
+        "confusion_matrix": confusion_matrix(ground_truths, predictions),
+    }
+
+
+def print_metrics(metrics: dict):
+    printo(f"Accuracy: {metrics['accuracy']}")
+    printo(f"Precision: {metrics['precision']}")
+    printo(f"Recall: {metrics['recall']}")
+    printo(f"F1: {metrics['f1']}")
+    printo(f"Confusion Matrix:\n{metrics['confusion_matrix']}")
+
+
+def get_groups_by_property(ids: list[str], property: str, benchmark: Benchmark):
+    groups = {}
+    for n, i in enumerate(ids):
+        if not isinstance(benchmark.data[i][property], list):
+            benchmark.data[i][property] = [benchmark.data[i][property]]
+        for j in benchmark.data[i][property]:
+            if j not in groups:
+                groups[j] = [n]
```

Comment:
Instantiate groups like this `groups = defaultdict(list[int])` and you don't need the `if j not in groups`.